Fix bandwidth related issues

guggero opened the pull request:
Depends on https://github.com/lightninglabs/taproot-assets/pull/1462.

Fixes https://github.com/lightninglabs/taproot-assets/issues/1471.
Fixes https://github.com/lightninglabs/taproot-assets/issues/1275.
Fixes https://github.com/lightninglabs/taproot-assets/issues/1374.

guggero commented:
Fixed unit test, ready for review.

ffranr commented on rfqmsg/custom_channel_data.go:
```diff
@@ -43,6 +48,26 @@ type JsonAssetChannel struct {
 
+// HasAllAssetIDs checks if the OpenChannel contains all asset IDs in the
+// provided set. It returns true if all asset IDs are present, false otherwise.
+func (c *JsonAssetChannel) HasAllAssetIDs(ids fn.Set[asset.ID]) bool {
+	for id := range ids {
+		found := false
+		for _, fundingAsset := range c.FundingAssets {
+			if fundingAsset.AssetGenesis.AssetID == id.String() {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return false
+		}
+	}
+
+	return true
+}
```
Could also construct two sets of id strings and then diff or union and length check.

ffranr commented on tapchannelmsg/records.go:
```diff
@@ -536,6 +556,42 @@ func (c *Commitment) Leaves() lnwallet.CommitAuxLeaves {
+		channelAssetIDs.Add(outgoing.AssetID.Val)
+	}
+
+	for id := range assetIDs {
+		found := false
+		for channelAssetID := range channelAssetIDs {
+			if channelAssetID == id {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return false
+		}
+	}
```
Set union or diff methods can probably be used here.

ffranr reviewed (approved):
Subject to George's comments, LGTM :+1: 

Could do with an itest in terminal I suppose to ensure that this fix remains effective. Perhaps there is already a PR for that?

guggero commented on tapchannel/aux_traffic_shaper.go:
```diff
@@ -81,27 +81,30 @@ func (s *AuxTrafficShaper) Stop() error {
-
-	// If there is no auxiliary blob in the channel, it's not a custom
-	// channel, and we don't need to handle it.
-	if fundingBlob.IsNone() {
```
But that's exactly the thing that was wrong. When an asset HTLC is routed over a non-asset channel, because of the randomization in channel selection, we _need_ to handle the traffic so we can return 0 bandwidth (meaning: DON'T use that channel to forward the HTLC).

guggero commented on tapchannel/aux_traffic_shaper.go:
```diff
@@ -149,6 +149,23 @@ func (s *AuxTrafficShaper) PaymentBandwidth(htlcBlob,
+			return b.AssetID.Val
+		})...,
+	)
+	if !commitment.HasAllAssetIDs(htlcAssetIDs) {
```
Yeah, you're right. I didn't consider this so things actually failed here. Working on a fix right now.

Roasbeef replied:
Perhaps once we land this fix, we should commit to just including the group key in all relevant areas for HTLCs/RFQ? Otherwise we'll always need to keep in mind the little trick to remember that sometimes an asset ID can _actually_ be a group key. Transmitting the group key within these wire messages and TLV blobs will also save us from needing to do a db/universe look up to verify that something that _looks_ like a group key is actually a group key.

guggero replied:
Yeah, I can attempt that in a follow-up PR. Though I think it might be quite a bit involved, but we'll see I guess.

Roasbeef commented on tapchannel/aux_invoice_manager.go:
```diff
@@ -415,6 +422,47 @@ func (s *AuxInvoiceManager) validateAssetHTLC(ctx context.Context,
+
+	// We also need to validate that the HTLC is actually the correct asset
+	// and arrived through the correct asset channel.
+	channels, err := s.cfg.LightningClient.ListChannels(ctx, true, false)
```
Hmm yeah was thinking something similar here. We don't actually store the custom chan data ourselves, so for now we must relay on `lnd` to deliver it to us. In other areas we pass it within the hook callbacks, but only the channels related to a given context.

Roasbeef replied:
IIUC, I think we can actually get rid of this extra RPC call:
  * With `handleInvoiceAccept` we're passed the actual invoice that we created. 
  * This invoice has an `scid` in the hop hint. 
  * We also have the circuit key here which contains the scid. 

Can't we parse out the hop hint earlier in the pipeline, pass it in here, then assert that it matches the circuit key scid? Would need to check if this is the scid alias or not passed in... 

Even with that though, we'd still the obtain the custom data from a channel to ensure that it can actually carry a given HTLC.

Roasbeef commented on tapchannel/aux_traffic_shaper.go:
```diff
@@ -115,33 +118,25 @@ func (s *AuxTrafficShaper) PaymentBandwidth(htlcBlob,
+	if len(commitmentBytes) == 0 {
+		log.Tracef("Empty commitment blob, cannot forward asset HTLC " +
+			"over non-asset channel, returning 0 bandwidth")
+		return 0, nil
```
👍

guggero commented on rfqmsg/custom_channel_data.go:
```diff
@@ -43,6 +48,26 @@ type JsonAssetChannel struct {
+// HasAllAssetIDs checks if the OpenChannel contains all asset IDs in the
+// provided set. It returns true if all asset IDs are present, false otherwise.
+func (c *JsonAssetChannel) HasAllAssetIDs(ids fn.Set[asset.ID]) bool {
+	for id := range ids {
```
Changed things to use set operations.

guggero commented:
Ready for re-review. Fixed all bugs and also updated the corresponding [integration tests](https://github.com/lightninglabs/lightning-terminal/pull/987).

The `litd` itests should also pass in the next re-try.
